Restore timing function

brad-decker opened the pull request:
<details>
<summary>Screenshot</summary>
<img src="https://user-images.githubusercontent.com/4448075/84531336-154ce480-acaa-11ea-86c3-5bd0e9aac401.png"/>


</details>

<details>
<summary>Screenshot INTL</summary>
<img src="https://user-images.githubusercontent.com/4448075/84531240-e6cf0980-aca9-11ea-95b5-5fdaa6498260.png" />


</details>

<details>
<summary>Screenshot Settings</summary>
<img src="https://user-images.githubusercontent.com/4448075/84531460-44635600-acaa-11ea-955a-a4b6b57f6abc.png" />


</details>


restores a feature that was erased by the transaction list refactor,
and greatly improves upon the UX. This new version is viewable in
both fullscreen and popup mode. The component was refactored into a
hook and is consumed in the transaction-list-item component now.

Note: I'm not sure where to actually activate this feature and I'm open
to adding a new toggle to the settings screen for this one. Unless I'm
just missing it.

Depends on #8745

brad-decker commented on ui/app/components/ui/icon/preloader/preloader-icon.component.js:
```diff
@@ -5,44 +5,48 @@ import classnames from 'classnames'
-      />
-    </g>
-  </svg>
+      <mask
```
None of this SVG actually changed, it just got shifted around a bit.

Gudahtt commented:
>Note: I'm not sure where to actually activate this feature and I'm open
to adding a new toggle to the settings screen for this one. Unless I'm
just missing it.

I don't think you're missing it - I can't see a toggle for this anywhere either! I guess adding one would be a good idea then.

Gudahtt commented on ui/app/hooks/useTransactionTimeRemaining.js:
```diff
@@ -0,0 +1,54 @@
+  const timeElapsedSinceSubmission = (currentTime - submittedTime) / 1000
+  const timeRemainingOnEstimate = initialTimeEstimate - timeElapsedSinceSubmission
+
+  const renderingTimeRemainingEstimate = `~ ${Math.round(timeRemainingOnEstimate / 60)}min`
```
This `min` string is not localized. I think this entire string will need to be localized in order to support other locales properly 🤔 

We could use this internationalization API instead: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat/format
But that would restrict us somewhat (we'd be stuck showing "1 min. ago", "this minute", or "in 1 min." for values of -1, 0, and 1 respectively, with no way to format it differently).

Gudahtt replied:
This API is pretty cool though! It'd be nice to automatically support all locales without any transaction effort.

Gudahtt replied:
I've discussed this with @rachelcope, and she agreed that the "narrow" relative time format would be OK to use here. It's not much longer.

Gudahtt replied:
We would need a polyfill though. Our minimum supported Chrome and Firefox versions are much older than the introduction of this API.

Gudahtt commented on ui/app/components/ui/icon/preloader/preloader-icon.component.js:
```diff
@@ -5,44 +5,48 @@ import classnames from 'classnames'
+        />
+      </g>
+    </svg>
+    {label && <span className="preloader__label">{label}</span>}
```
It seems a bit strange to me to include a label with the icon component directly. I would expect this component to be _just_ the icon, without anything accompanying it. I believe that's what the others are as well.

Could this label be extracted and put in `TransactionListItem` instead?

brad-decker replied:
I thought maybe it might be something that could be useful elsewhere as well, so I created a new component to do this called IconWithLabel. Lemme know if that works better or if you'd still like to see it in `TransactionListItem`

Gudahtt replied:
`IconWithLabel` sounds like a great idea to me!

brad-decker committed:
restore and enhance the time est feature

background: we had a feature for showing a time estimate on pending txs
that was accidently removed during the redesign implementation. This PR
restores that feature and also enhances it:
1. Displays the time estimate on all views instead of just fullscreen
2. Uses Intl.RelativeTimeFormat to format the time
3. Adds a way to toggle the feature flag.
4. Uses a hook to calculate the time remaining instead of a component

Gudahtt commented on ui/app/components/app/transaction-list-item/transaction-list-item.component.js:
```diff
@@ -22,6 +22,8 @@ import {
 import TransactionStatus from '../transaction-status/transaction-status.component'
 import TransactionIcon from '../transaction-icon'
+import { useTransactionTimeRemaining } from '../../../hooks/useTransactionTimeRemaining'
+import IconWithLabel from '../../ui/icon-with-label/icon-with-label'
```
I noticed that you did add an `index.js` file for `icon-with-label`, but the file appears to be empty :thinking:

brad-decker replied:
oops

brad-decker replied:
resolved

Gudahtt commented on ui/app/hooks/useTransactionTimeRemaining.js:
```diff
@@ -0,0 +1,92 @@
+  const featureFlags = useSelector(getFeatureFlags)
+  const transactionTimeFeatureActive = featureFlags?.transactionTime
+
+  const rtf = new Intl.RelativeTimeFormat(locale, { style: 'narrow' })
```
It might be better to use `numeric: 'auto'`, so that <1 minute is shown as "this minute" instead of `in 0 min.`

```suggestion
  const rtf = new Intl.RelativeTimeFormat(locale, { numeric: 'auto', style: 'narrow' })
```

brad-decker replied:
Good call

brad-decker replied:
done

Gudahtt reviewed (dismissed):
LGTM!

brad-decker commented:
I actually just realized that there is nothing preventing this from working on testnets. I'm going to add that now.

Gudahtt commented:
Ah, right, good catch!

Gudahtt commented on app/_locales/en/messages.json:
```diff
@@ -1600,6 +1600,9 @@
     "message": "Transaction Time"
   },
+  "showTransactionTimeDescription": {
+    "message": "Select this to display pending transaction time estimates in the activity tab. Note: estimates are approximations based on network conditions."
```
Perhaps we should specify here that it's specific to mainnet as well?
```suggestion
    "message": "Select this to display pending transaction time estimates in the activity tab while on the Main Ethereum Network. Note: estimates are approximations based on network conditions."
```

brad-decker replied:
Thanks!

brad-decker committed:
Update app/_locales/en/messages.json

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

brad-decker committed:
do not display on test nets

brad-decker committed:
Merge branch 'restore-timing-function' of https://github.com/MetaMask/metamask-extension into restore-timing-function

brad-decker commented:
added the logic for not displaying outside of mainnet and included your suggestion to language update, @Gudahtt
